source/common/json/json_streamer.h

```diff
@@ -79,15 +136,16 @@ class StringOutput {
  * add(absl::string_view) and
  * add(absl::string_view, absl::string_view, absl::string_view) methods.
  */
-template <class OutputBufferType> class StreamerBase {
+template <class OutputBufferType> class StreamerBase : public SerializerBase<OutputBufferType> {
```

Comment:
I like how this shares some code, but I'm not crazy about how it exposes methods to JsonStreamer that break its abstraction boundary. You can get invalid Json out of the streamer by calling a public method on its base class.

There are a few ways to work around this, but I think the one that makes the most sense to me is to leave the `add*Delimeter` methods in the StringSerializer and not use them at all in the streamer. I don't think they add much readability and the abstraction is no more compact than the code they encapsulate.

I'll also note: the only code that is shared that isn't just a single line of impl per method is addNumber.

I'm not sure this co-mingling is worth it. WDYT?

Reply:
I would be fine with a templatized 'addNumber' free function also. That might be a lot easier.

Reply:
> I'll also note: the only code that is shared that isn't just a single line of impl per method is addNumber.

I think make the same logic in same position still make sense. And I do reduce lots of code lines.

Reply:
Regarding line-counts, json_streamer.h and .cc did not appear to shrink much with this change. I think this PR mostly moves code out of the serializer and into the shared base. I don't see the line-count benefit, but I do see the broadening of the interface for the Streamer to enable invalid json to be generated using legal calls.

I think if you boil this down the main thing to share is addNumber. The other thing to share was string sanitization but that is already done via the Sanitizer class.

Reply:
> I do see the broadening of the interface for the Streamer to enable invalid json to be generated using legal calls.

will this be a problem even after we moved the add*delimiter out?

Reply:
idk. It just seems awkward to re-inherit. It could just be a free function helper like we have for sanitization, or like we have for converting the floats to strings (without the NaN -> null)

Reply:
Considering most of the method are one line methods. Then I think it may make no big sense? I will hung this up and close this PR first. it doesn't matter anyway. :)